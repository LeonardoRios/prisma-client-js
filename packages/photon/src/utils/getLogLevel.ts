type LogLevel = 'info' | 'query' | 'warn'
type LogDefinition = {
  level: LogLevel
  emit: 'stdout' | 'event'
}

export function getLogLevel(
  log: Array<LogLevel | LogDefinition>,
): LogLevel | undefined {
  return log.reduce<LogLevel | undefined>((acc, curr) => {
    const currentLevel = typeof curr === 'string' ? curr : curr.level
    if (currentLevel === 'query') {
      return acc
    }
    if (!acc) {
      return currentLevel
    }
    if (curr === 'info' || acc === 'info') {
      // info has precedence
      return 'info'
    }
    return currentLevel
  }, undefined)
}
