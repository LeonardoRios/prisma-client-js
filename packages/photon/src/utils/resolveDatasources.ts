import { DataSource } from '@prisma/generator-helper'
import { uriToCredentials } from '@prisma/sdk/src/convertCredentials'
import path from 'path'
import { credentialsToUri } from '@prisma/sdk'

export function resolveDatasources(datasources: DataSource[], cwd: string, outputDir: string): DataSource[] {
  return datasources.map(datasource => {
    if (datasource.connectorType === 'sqlite') {
      if (datasource.url.fromEnvVar === null) {
        return {
          ...datasource,
          url: {
            fromEnvVar: null,
            value: absolutizeSqliteRelativePath(datasource.url.value, cwd, outputDir),
          },
        }
      } else {
        return datasource
      }
    } else if (datasource.connectorType === 'postgresql') {
      return {
        ...datasource,
        url: {
          value: absolutizePostgreSQLRelativePath(datasource.url.value, cwd, outputDir),
        },
      }
    }
    return datasource
  })
}

export function absolutizeSqliteRelativePath(url: string, cwd: string, outputDir: string): string {
  let filePath = url

  if (filePath.startsWith('file:')) {
    filePath = filePath.slice(5)
  }

  const absoluteTarget = path.resolve(cwd, filePath)

  return `'file:' + path.resolve(__dirname, '${path.relative(outputDir, absoluteTarget)}')`
}

export function absolutizePostgreSQLRelativePath(url: string, cwd: string, outputDir: string): string {
  const credentials = uriToCredentials(url)
  
  if (credentials.extraFields?.sslcert) {
    const absoluteTarget = path.resolve(cwd, credentials.extraFields.sslcert)
    credentials.extraFields.sslcert = path.relative(outputDir, absoluteTarget)
  }

  if (credentials.extraFields?.sslidentity) {
    const absoluteTarget = path.resolve(cwd, credentials.extraFields.sslidentity)
    credentials.extraFields.sslidentity = path.relative(outputDir, absoluteTarget)
  }

  return credentialsToUri(credentials)
}