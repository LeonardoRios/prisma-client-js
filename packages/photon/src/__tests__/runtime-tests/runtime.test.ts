import fs from 'fs'
import path from 'path'
import { generateInFolder } from '../../utils/generateInFolder'

jest.setTimeout(30000)

describe('runtime works', () => {
  const subDirs = getSubDirs(__dirname)
  for (const dir of subDirs) {
    const testName = path.basename(dir)

    test(`can run ${testName} example`, async () => {
      const envVars = getEnvVars(dir)
      process.env = { ...process.env, ...envVars }

      await generateInFolder({
        projectDir: dir,
        useLocalRuntime: false,
        transpile: true,
      })

      if (envVars) {
        for (const key of Object.keys(envVars)) {
          delete process.env[key]
        }
      }

      const filePath = path.join(dir, 'index.js')
      const fn = require(filePath)

      if (envVars) {
        try {
          await fn()
        } catch (e) {
          expect(e).toMatchSnapshot()
        }
      } else {
        expect(fn()).resolves.toMatchSnapshot()
      }
    })
  }
})

function getSubDirs(dir: string): string[] {
  const files = fs.readdirSync(dir)
  return files
    .map(file => path.join(dir, file))
    .filter(
      file =>
        fs.lstatSync(file).isDirectory() && !file.includes('__snapshots__'),
    )
}

function getEnvVars(dir: string): { [key: string]: string } | undefined {
  const envPath = path.join(dir, 'env.json')
  if (fs.existsSync(envPath)) {
    return JSON.parse(fs.readFileSync(envPath, 'utf-8'))
  }

  return undefined
}
