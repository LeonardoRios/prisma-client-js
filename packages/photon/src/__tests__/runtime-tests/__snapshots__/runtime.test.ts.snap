// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime works blog example should succeed: blog example should succeed 1`] = `undefined`;

exports[`runtime works blog-env example should not succeed: blog-env example should not succeed 1`] = `
[Error: error: Environment variable not found: SQLITE_URL.
  -->  schema.prisma:3
   | 
 2 |   provider = "sqlite"
 3 |   url      = env("SQLITE_URL")
   | 

Validation Error Count: 1]
`;

exports[`runtime works enums example should succeed: enums example should succeed 1`] = `undefined`;
